[Release 3.17821.0] COE Review (#150)
cloudbees/terraform-aws-cloudbees-ci-eks-addon: CONTRIBUTING.md

## Design principles

- It follows the same approach as the [Terraform AWS EKS Blueprints for Terraform Patterns](https://aws-ia.github.io/terraform-aws-eks-blueprints/).

> - Blueprint uses a Monorepo configuration: Required external repositories for configuration are included within the same repository. For example, the at scale blueprint contains the repository for CasC bundles.

Reviewer: ```suggestion
- The blueprints use a monorepo configuration; the required external repositories for the configuration are included within the same repository. For example, the At scale blueprint contains the repository for CasC bundles.
```

> - Submit pull request against `develop` branch and release from `main` branch.

Reviewer: ```suggestion
- Submit pull requests against the `develop` branch and release from the `main` branch.
```

>   - `main`:

Reviewer: ```suggestion
  - `main` branch:
```

>     - It is the stable branch and it is used for releases.
>     - Before merging a pull request, the CloudBees Center of Excellence (CoE) team must validate the changes.
>     - Requisites:
>       - The field `source` in the `eks_blueprints_addon_cbci` at blueprints must point to the remote [terraform registry version](https://registry.terraform.io/modules/cloudbees/cloudbees-ci-eks-addon/aws/latest) and `version >= "x.x.x"`. It is important for the telemetry in https://registry.terraform.io/modules/cloudbees/cloudbees-ci-eks-addon/aws/latest.
>       - The CasC bundles SCM configuration must point to the repository https://github.com/cloudbees/terraform-aws-cloudbees-ci-eks-addon and its `main` branch.
>   - `develop`:
>     - It is the integration branch and it is used for testing new features and updates before merging them into the `main` branch.
>     - Requisites:
>       - The field `source` in the `eks_blueprints_addon_cbci` in the blueprints folder must point to the local CloudBees CI Terraform EKS Addon root of the repository `source = "../../"`.
>       - The CasC bundles SCM configuration must point to the repository https://github.com/cloudbees/terraform-aws-cloudbees-ci-eks-addon and its `develop` branch.

Reviewer: ```suggestion
    - It is the stable branch and is used for releases.
    - Before merging a pull request, the CloudBees Center of Excellence (CoE) team must validate the changes.
    - Requirements:
      - The `source` field in the `eks_blueprints_addon_cbci` at blueprints must point to the remote [terraform registry version](https://registry.terraform.io/modules/cloudbees/cloudbees-ci-eks-addon/aws/latest) and `version >= "x.x.x"`. It is important for the telemetry in https://registry.terraform.io/modules/cloudbees/cloudbees-ci-eks-addon/aws/latest.
      - The CasC bundles SCM configuration must point to the https://github.com/cloudbees/terraform-aws-cloudbees-ci-eks-addon repository and its `main` branch.
  - `develop` branch:
    - It is the integration branch and it is used for testing new features and updates before merging them into the `main` branch.
    - Requirements:
      - The `source` field in the `eks_blueprints_addon_cbci` in the blueprints folder must point to the local root of the https://github.com/cloudbees/terraform-aws-cloudbees-ci-eks-addon repository (for example, `source = "../../"`).
      - The CasC bundles SCM configuration must point to the https://github.com/cloudbees/terraform-aws-cloudbees-ci-eks-addon repository and its `develop` branch.
```

Contributions via pull requests are appreciated. Before submitting a pull request, please ensure that you:

1. Are working against the latest source on the `develop` branch.
2. Check existing open, and recently merged, pull requests to make sure someone else has not already addressed the problem.
3. Open an issue to discuss any significant work; we do not want your time to be wasted.

To submit a pull request:

1. Fork the repository.

> 2. Create a feature branch extending from the `develop` branch.

Reviewer: ```suggestion
2. Create a feature branch based on the `develop` branch.
```

> 3. Modify the source and focus on the specific change you are contributing. For example, if you reformat all the code, it is hard for reviewers to focus on your specific change.

Reviewer: ```suggestion
3. Modify the source files and focus on the specific change you are contributing. For example, if you reformat all of the code, it is difficult for reviewers to focus on your specific change.
```

4. **Ensure that local tests pass**. Local tests can be orchestrated via the companion [Makefile](Makefile).
5. Make commits to your fork using clear commit messages.
6. Submit a pull request against the `develop` branch and answer any default questions in the pull request interface.
7. Pay attention to any automated CI failures reported in the pull request, and stay involved in the conversation.

> [!IMPORTANT]

> > If you make updates to embeded repository (e.g. CasC bundles), you must push the changes to the public upstream (repository/branch) before running `terraform apply` locally. The endpoint and/or branch can be updated via `set-casc-location` from the companion [Makefile](Makefile).

Reviewer: ```suggestion
> If you make updates to an embedded repository (for example, CasC bundles), you must push the changes to the public upstream (repository/branch) before running `terraform apply` locally. The endpoint and/or branch can be updated via the `set-casc-location` command from the companion [Makefile](Makefile).
```

The [bp-tf-ci.yaml](.cloudbees/workflows/bp-tf-ci.yaml) blueprints are orchestrated into the [CloudBees platform](https://www.cloudbees.com/products/saas-platform) inside the [CloudBees Professional Services (PS) sub-organization](https://cloudbees.io/orgs/cloudbees~professional-services/components/94c50dcf-125e-4767-b9c5-58d6d669a1f6/runs).

> [!NOTE]

> > At the time of writing, the pipeline triggers on `push` events only and not for `pull_requests`. Although pull request event is supported, it is requires filters for file patters ()`*.tf`).

Reviewer: ```suggestion
> Currently, the pipeline triggers on `push` events only and not for `pull_requests`. Although the pull request event is supported, it is requires filters for file patterns (for example, `*.tf`).
```

CloudBees CI Terraform EKS Addon versions try to be in sync with the [CloudBees CI releases](https://docs.cloudbees.com/docs/release-notes/latest/cloudbees-ci/).

> 1. Ensure that `develop` branch follows its requisites from the [Design principles](#design-principles) section.
> 2. Test locally the (`develop`) for all the blueprints. Use the `test-all` target in the companion [Makefile](Makefile).
> 3. Once all local tests passed successfully, create a PR against the `main` branch. **It requires to pass the COE Team validation**.
> 4. Once the pull request is merged, update the `main` branch following its requisites from the [Design principles](#design-principles) section. The [Blueprint Terraform CI pipeline](#blueprint-terraform-ci-pipeline) must validate the changes.
> 5. Create a [new release](https://github.com/cloudbees/terraform-aws-cloudbees-ci-eks-addon/releases). The release version semantics follow the Helm chart convention.

Reviewer: ```suggestion
1. Ensure that the `develop` branch follows the requirements from the [Design principles](#design-principles) section.
2. Locally test the `develop` branch for all the blueprints, using the `test-all` target in the companion [Makefile](Makefile).
3. Once all local tests have passed successfully, create a PR against the `main` branch. **It must pass for the CoE team validation**.
4. Once the pull request is merged, update the `main` branch following the requirements from the [Design principles](#design-principles) section. The [Blueprint Terraform CI pipeline](#blueprint-terraform-ci-pipeline) must validate the changes.
5. Create a [new release](https://github.com/cloudbees/terraform-aws-cloudbees-ci-eks-addon/releases). The release version semantics follow the Helm chart conventions.
```
